app/gui2/src/components/GraphEditor/GraphNodes.vue

```diff
@@ -1,15 +1,96 @@
 import { injectGraphSelection } from '@/providers/graphSelection'
 import { useGraphStore } from '@/stores/graph'
-import type { Vec2 } from '@/util/vec2'
+import { useApproach } from '@/util/animation'
+import { Rect } from '@/util/rect'
+import { Vec2 } from '@/util/vec2'
+import { iteratorFilter } from 'lib0/iterator'
+import { abs } from 'lib0/math'
 import type { ContentRange, ExprId } from 'shared/yjsModel'
+import { ref, watchEffect, type Ref, type WatchStopHandle } from 'vue'
 
 const graphStore = useGraphStore()
 const selection = injectGraphSelection(true)
 const navigator = injectGraphNavigator(true)
 
+class Drag {
+  static THRESHOLD = 15.0
+
+  draggedNodes: ExprId[]
+  grid: SnapGrid
+  offset = ref(Vec2.Zero)
+  snapXTarget = ref(0.0)
+  snapX = useApproach(this.snapXTarget, 30.0)
+  snapYTarget: Ref<number> = ref(0.0)
+  snapY = useApproach(this.snapYTarget, 30.0)
+  stopPositionUpdate: WatchStopHandle
+
+  constructor(movedId: ExprId) {
+    this.draggedNodes = selection?.isSelected(movedId) ? Array.from(selection.selected) : [movedId]
+    this.grid = this.createSnapGrid()
+
+    this.stopPositionUpdate = watchEffect(() => {
+      for (const id of this.draggedNodes) {
+        const node = graphStore.nodes.get(id)
+        if (node == null) continue
+        node.visiblePosition = node.position
+          .add(this.offset.value)
+          .add(new Vec2(this.snapX.value, this.snapY.value))
+      }
+    })
+  }
+
+  updateOffset(offset: Vec2): void {
+    const oldSnappedOffset = this.offset.value.add(new Vec2(this.snapX.value, this.snapY.value))
+    this.offset.value = offset
+    let rects: Rect[] = []
+    for (const id of this.draggedNodes) {
+      const rect = graphStore.nodeRects.get(id)
+      const node = graphStore.nodes.get(id)
+      if (rect != null && node != null) rects.push(new Rect(node.position.add(offset), rect.size))
+    }
+    const snap = this.grid.snappedMany(rects, Drag.THRESHOLD)
+    const newSnappedOffset = offset.add(snap)
+    this.snapXTarget.value = snap.x
+    if (abs(newSnappedOffset.x - oldSnappedOffset.x) < 2.0) {
+      this.snapX.skip()
+    }
+    this.snapYTarget.value = snap.y
+    if (abs(newSnappedOffset.y - oldSnappedOffset.y) < 2.0) {
+      this.snapY.skip()
+    }
+  }
+
+  finishDragging(): void {
+    this.stopPositionUpdate()
+    for (const id of this.draggedNodes) {
+      const node = graphStore.nodes.get(id)
+      if (node == null || node.visiblePosition == null) continue
+      const newPosition = node.position
+        .add(this.offset.value)
+        .add(new Vec2(this.snapXTarget.value, this.snapYTarget.value))
+      console.log('New Position', newPosition)
+      graphStore.setNodePosition(id, newPosition)
+      delete node.visiblePosition
+    }
+  }
+
+  private createSnapGrid() {
+    const excludeSet = new Set<ExprId>()
+    for (const node of this.draggedNodes) excludeSet.add(node)
+    const withoutExcluded = iteratorFilter(
```

Comment:
instead of `iteratorFilter` + `Array.from(, mapFn)` you could perhaps use a generator function to do the filter and the map at the same time, and then `return new SnapGrid(Array.from(this.getSnapTargets()))` ,for example

Reply:
Yes, I could, but why it is better? I think using `iteratorFilter` instead of generator function is more compact, and thus better express the intention (and `Array.from` is used anyway).